### PR TITLE
ionctl & management api & end2end integration work

### PR DESCRIPTION
**Pull request by lawrencegripper**
related #122 

Edit: This has become a bit of a monster PR to integrate the work form @jjcollinge, me and @JLugagne so that from the frontapi -> a successfully run module works end2end.

### CODE REVIEW

File: cmd/ionctl/root/root.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+		// Find home directory.
+		home, err := homedir.Dir()
+		if err != nil {
+			fmt.Println(err)
+			os.Exit(1)
+		}
+
+		// Search config in home directory with name ".ioncli" (without extension).
+		viper.AddConfigPath(home)
+		viper.SetConfigName(".ioncli")
```
Comment (jjcollinge) [inline cmd/ionctl/root/root.go]: we'll need to be consistent `.ioncli` or `.ionctl`?

Comment (lawrencegripper) [reply cmd/ionctl/root/root.go]: Yeah good point, I flip flopped on the naming a bit. Do you think it's better to go as ionctl or ioncli?

Comment (lawrencegripper) [reply cmd/ionctl/root/root.go]: Or I just call the cli tool `ion` to make things either simpler

**Comment by lawrencegripper** [comment]
Gonna fix up the linting issues now

File: cmd/ionctl/root/root.go (hunk 1 above)
Comment (jjcollinge) [reply cmd/ionctl/root/root.go]: I'd rather go with brevity for a CLI - `ion` is better. We'd end up aliasing `ionctl` to `i` or `ion` anyway. We can refer to it as ion CLI in docs etc as it's clearer. We don't want the `kubectl` debate on how the hell you say it!

Comment (lawrencegripper) [reply cmd/ionctl/root/root.go]: Yeah that makes sense - lets got for `ion`

**Comment by lawrencegripper** [comment]
So I think this is very nearly good to go, just need to fix up #137 then we're all set.

File: cmd/ion/dev/listevents.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	for _, file := range files {
+		b, err := ioutil.ReadFile(file)
+		if err != nil {
+			return fmt.Errorf("error reading module event %s: %+v", file, err)
+		}
+		var event common.Event
+		if err = json.Unmarshal(b, &event); err != nil {
+			return fmt.Errorf("error parsing module event %s: %+v", file, err)
+		}
+		fmt.Println(string(b))
```
[the remark this answers is not in the document]
Comment (lawrencegripper) [reply cmd/ion/dev/listevents.go]: I'll leave this one for @jjcollinge as I merged in some WIP while he was away, sorry joni!

File: internal/app/dispatcher/run.go [hunk 3]
```diff
@@ -55,27 +55,28 @@ func Run(cfg *types.Configuration) {
...
-				log.WithError(err).Error("Couldn't dispatch message to kubernetes provider")
+				contextualLogger.WithError(err).Error("Couldn't dispatch message to kubernetes provider")
 			}
 
-			log.WithField("message", message).Debug("message dispatched")
-			queueDepth, err := amqpConnection.GetQueueDepth()
+			contextualLogger.Debug("message dispatched")
+			queueStats, err := amqpConnection.GetQueueDepth()
 			if err != nil {
 				log.WithError(err).Error("failed getting queue depth from listener")
```
[the remark this answers is not in the document]
Comment (lawrencegripper) [reply internal/app/dispatcher/run.go]: good spot, will fix up

File: cmd/frontapi/serve.go [hunk 4]
```diff
@@ -16,40 +16,59 @@ import (
...
+	flags.BoolP("printconfig", "P", false, "Print out config when starting")
+
 	// job.*
-	serveCmd.PersistentFlags().Int("job.retrycount", 0, "Max number of times a job can be retried")
+	flags.Int("job.retrycount", 0, "Max number of times a job can be retried")
+
+	// document store flags
+	flags.String("mongodb-name", "", "MongoDB Name")
+	flags.String("mongodb-collection", "", "MongoDB Database Collection")
+	flags.String("mongodb-username", "", "MongoDB server username")
```
[the remark this answers is not in the document]
Comment (lawrencegripper) [reply cmd/frontapi/serve.go]: ta fixed

File: cmd/frontapi/serve.go [hunk 5]
```diff
@@ -16,40 +16,59 @@ import (
...
-	serveCmd.PersistentFlags().String("moduleconfigpath", "", "Path to environment variables file for module")
-	serveCmd.PersistentFlags().BoolP("printconfig", "P", false, "Print out config when starting")
+	flags.StringVarP(&cfgFile, "config", "c", "../../configs/frontapi.yaml", "Config file path")
+	flags.StringP("loglevel", "l", "warn", "Log level (debug|info|warn|error)")
+	flags.String("modulename", "", "Name of the module")
+	flags.String("subscribestoevent", "", "Event this modules subscribes to")
+	flags.String("eventspublished", "", "Events this modules can publish")
+	flags.String("servicebusnamespace", "", "Namespace to use for ServiceBus")
+	flags.String("resourcegroup", "", "Azure ResourceGroup to use")
+	flags.Bool("logsensitiveconfig", false, "Print out sensitive config when logging")
```
[the remark this answers is not in the document]
Comment (lawrencegripper) [reply cmd/frontapi/serve.go]: ta fixed

File: cmd/frontapi/serve.go [hunk 6]
```diff
@@ -16,40 +16,59 @@ import (
...
+	flags.String("moduleconfigpath", "", "Path to environment variables file for module")
```
[the remark this answers is not in the document]
Comment (lawrencegripper) [reply cmd/frontapi/serve.go]: ta fixed

File: internal/app/frontapi/integration/integration_test.go [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+		}
+
+		client := http.Client{}
+		res, err := client.Do(request)
+		if err != nil {
+			t.Log(err)
+			continue
+		}
+
+		resBytes, _ := ioutil.ReadAll(res.Body)
```
[the remark this answers is not in the document]
Comment (lawrencegripper) [reply internal/app/frontapi/integration/integration_test.go]: Great spot didn't realize that, had the same bug in the azure.go file in handler data plane which I've also fixed up.

File: internal/app/handler/integration/azure_intergration_test.go [hunk 8]
```diff
@@ -153,6 +169,10 @@ func TestAzureIntegration(t *testing.T) {
 		t.Fatalf("error decoding file '%s' content: '%+v'", environment.InputMetaFilePath, err)
 	}
 
+	if len(kvps) != 1 {
+		t.Fatalf("insights file should contain 1 key value pair, but has %d", len(kvps))
+	}
+
 	// The first key, value pair should be as expected
 	for _, kvp := range kvps {
```
[the remark this answers is not in the document]
Comment (lawrencegripper) [reply internal/app/handler/integration/azure_intergration_test.go]: Agree, removed.

File: internal/app/management/client/client.go [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+	}
+	for _, name := range listResponse.Names {
+		fmt.Printf("%s\n", name)
+	}
+
+	time.Sleep(5 * time.Second)
+
+	// Get module
+	moduleIsAvailable := false
+	for !moduleIsAvailable {
```
[the remark this answers is not in the document]
Comment (lawrencegripper) [reply internal/app/management/client/client.go]: This was a file used while testing the GRPC API, not the management cli is up and running I think best to remove it entirely.

File: internal/pkg/common/event.go [hunk 10]
```diff
@@ -13,32 +13,47 @@ type KeyValuePair struct {
...
-func (kvps KeyValuePairs) Append(kvp KeyValuePair) {
-	kvps = append(kvps, kvp)
+func (kvps KeyValuePairs) Append(kvp KeyValuePair) KeyValuePairs {
+	return append(kvps, kvp)
 }
 
 //Remove a key value pair at an index by shifting the slice
-func (kvps KeyValuePairs) Remove(index int) error {
+func (kvps KeyValuePairs) Remove(index int) (KeyValuePairs, error) {
 	if (index > len(kvps)+1) || (index < 0) {
```
[the remark this answers is not in the document]
Comment (lawrencegripper) [reply internal/pkg/common/event.go]: Good spot I've fixed this up.

File: cmd/ion/dev/listmodules.go [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+	if err != nil {
+		return fmt.Errorf("error finding module directories in %s: %+v", ionModulesDir, err)
+	}
+	if len(files) == 0 {
+		fmt.Println("no modules available")
+		return nil
+	}
+
+	for _, file := range files {
+		file := strings.Replace(file, ionModulesDir, "", -1)
```
[the remark this answers is not in the document]
Comment (lawrencegripper) [reply cmd/ion/dev/listmodules.go]: Agreed I'll change that, thanks.

File: cmd/ion/dev/listevents.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jjcollinge) [reply cmd/ion/dev/listevents.go]: Yeh the 'dev' module in the CLI is full of WIP - this was just for debug to see the event - having formatters would be nicer - something we should flag to fix later.

File: internal/app/frontapi/integration/integration_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (lawrencegripper) [reply internal/app/frontapi/integration/integration_test.go]: Hasn't picked up the change as it's below the line :( have added 

```
		resBytes, _ := ioutil.ReadAll(res.Body)
		defer res.Body.Close() //nolint: errcheck
```

File: cmd/ion/event/get.go [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+var getCmd = &cobra.Command{
+	Use:   "get",
+	Short: "get an event from the messaging queue (dequeue)",
+	RunE:  Get,
+}
+
+// Get an event from the ion eventing system (dequeues the message)
+func Get(cmd *cobra.Command, args []string) error {
+
+	subPath := fmt.Sprintf("/%s/subscriptions/%s", getOpts.eventType, getOpts.subName)
```
[the remark this answers is not in the document]
Comment (lawrencegripper) [reply cmd/ion/event/get.go]: This is building up a url so didn't seem appropriate

File: cmd/ion/event/get.go [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+	}()
+
+	// Receive next event
+	msg, err := receiver.Receive(ctx)
+	if err != nil {
+		return fmt.Errorf("error reading event: %+v", err)
+	}
+
+	var event common.Event
+	b := msg.GetData()
```
[the remark this answers is not in the document]
Comment (lawrencegripper) [reply cmd/ion/event/get.go]: yup will fix this up

File: cmd/ion/event/get.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (JLugagne) [reply cmd/ion/event/get.go]: ok, nevermind then, I miss-understood this part

File: cmd/ion/event/peek.go [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+var peekCmd = &cobra.Command{
+	Use:   "peek",
+	Short: "peek an event from the messaging queue",
+	RunE:  Peek,
+}
+
+// Peek takes the latest event of the queue but does not remove it
+func Peek(cmd *cobra.Command, args []string) error {
+
+	subPath := fmt.Sprintf("/%s/subscriptions/%s", peekOpts.eventType, peekOpts.subName)
```
[the remark this answers is not in the document]
Comment (lawrencegripper) [reply cmd/ion/event/peek.go]: Yeah again it's a url so seemed not to be the right fit

File: internal/app/dispatcher/providers/kubernetes.go [hunk 15]
```diff
@@ -343,14 +330,20 @@ func (k *Kubernetes) Dispatch(message messaging.Message) error {
...
+
 func homeDir() string {
 	if h := os.Getenv("HOME"); h != "" {
 		return h
 	}
 	return os.Getenv("USERPROFILE") // windows
 }
 
-func getStatsFromJob(job *batchv1.Job) *log.Entry {
+func getLoggerForJob(job *batchv1.Job) *log.Entry {
```
[the remark this answers is not in the document]
Comment (lawrencegripper) [reply internal/app/dispatcher/providers/kubernetes.go]: Great spot, fixed this up now

**Pull request merged**